opentelemetry/src/runtime.rs

```diff
@@ -0,0 +1,122 @@
+///
+/// [Tokio]: https://crates.io/crates/tokio
+/// [async-std]: https://crates.io/crates/async-std
+pub trait Runtime {
+    /// A future stream, which returns items in a previously specified interval. The item type is
+    /// not important.
+    type Interval: Stream + Send;
+
+    /// A future, which resolves after a previously specified amount of time. The output type is
+    /// not important.
+    type Delay: Future + Send;
+
+    /// Create a [Stream][futures::Stream], which returns a new item every
+    /// [Duration][std::time::Duration].
+    fn interval(&self, duration: Duration) -> Self::Interval;
+
+    /// Spawn a new task or thread, which executes the given future.
+    fn spawn(&self, future: BoxFuture<'static, ()>);
+
+    /// Return a new future, which resolves after the specified [Duration][std::time::Duration].
+    fn delay(&self, duration: Duration) -> Self::Delay;
+}
+
+/// Runtime implementation, which works with Tokio's multi thread runtime.
+#[cfg(feature = "rt-tokio")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rt-tokio")))]
+#[derive(Debug, Clone)]
+pub struct Tokio;
+
+#[cfg(feature = "rt-tokio")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rt-tokio")))]
+impl Runtime for Tokio {
+    type Interval = tokio_stream::wrappers::IntervalStream;
+    type Delay = tokio::time::Sleep;
+
+    fn interval(&self, duration: Duration) -> Self::Interval {
+        crate::util::tokio_interval_stream(duration)
+    }
+
+    fn spawn(&self, future: BoxFuture<'static, ()>) {
+        let _ = tokio::spawn(future);
+    }
+
+    fn delay(&self, duration: Duration) -> Self::Delay {
+        tokio::time::sleep(duration)
+    }
+}
+
+/// Runtime implementation, which works with Tokio's current thread runtime.
+#[cfg(feature = "rt-tokio-current-thread")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rt-tokio-current-thread")))]
+#[derive(Debug, Clone)]
+pub struct TokioCurrentThread;
+
+#[cfg(feature = "rt-tokio-current-thread")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rt-tokio-current-thread")))]
+impl Runtime for TokioCurrentThread {
+    type Interval = tokio_stream::wrappers::IntervalStream;
+    type Delay = tokio::time::Sleep;
+
+    fn interval(&self, duration: Duration) -> Self::Interval {
+        crate::util::tokio_interval_stream(duration)
+    }
+
+    fn spawn(&self, future: BoxFuture<'static, ()>) {
+        // We cannot force push tracing in current thread tokio scheduler because we rely on
+        // BatchSpanProcessor to export spans in a background task, meanwhile we need to block the
+        // shutdown function so that the runtime will not finish the blocked task and kill any
+        // remaining tasks. But there is only one thread to run task, so it's a deadlock
+        //
+        // Thus, we spawn the background task in a separate thread.
+        std::thread::spawn(move || {
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap();
```

Comment:
I was reviewing this part the other day and thought I probably should have added a `expect` here just in case the build is failed. Could we add something like this?

Reply:
Yes, that sounds a good idea. 👍  I added this change in this PR.